### PR TITLE
exp/lighthorizon: Add a set of tools to aide in index inspection.

### PR DESCRIPTION
**Pull request by Shaptic**
### What
This extends the `tools` toolkit (which previously only let you inspect the web server ledger cache) to add helpful tools to inspect indexes. It includes the following tools: when given an index location, you can

 * view what accounts are indexed there
 * view what checkpoints particular accounts are active at
 * view statistics about the index such as:
   - the # of accounts indexed, 
   - the range (smallest, largest) of indexed checkpoints
   - the total # of active checkpoints (thus giving you a ledger range covered by the index)
   - the index names

### Why
This is particularly useful until we extend the index format to store some of this stuff internally, so that we can learn about long-running processes (like the map-reduce index builder) and how much progress they've made.

For example, I used this toolset to learn that there was ~1mo of indexed data on S3 for the demo.

### Known limitations
n/a

### Examples

Command: `./tools index view file://$HOME/workspace/indices --limit 5`
Output:

```
INFO[2022-08-29T15:47:15.823-07:00] Analyzing indices at file:///home/george/workspace/indices  pid=241771
INFO[2022-08-29T15:47:15.836-07:00] GCJ7F2CINZQ2WWXTS2X3BOY4UUAM4X7Q4PSKCQPI6URA56AQAUND724S  pid=241771
INFO[2022-08-29T15:47:15.836-07:00] GCMY3A5MTQKSFQYMJWTNVVMSJ3GRW567UVPIP4TYZGJ2O5YGCLIPVUQ7  pid=241771
INFO[2022-08-29T15:47:15.836-07:00] GA2HILXX64DMVBRJCRRXYUFNB5DKT5UOTSIY7RCHTDVKLHL2IARY7U52  pid=241771
INFO[2022-08-29T15:47:15.836-07:00] GAQVDLMFY4M64PHLB3U4Q6QLESTJWTMUVBVJA4H5DIFFPHMPSZAYGBI6  pid=241771
INFO[2022-08-29T15:47:15.836-07:00] GC2TQLMT3YJCC24VCYM6ZP46WWPXE22IVP4DEPNFHEQDJ3MZ7QTNFD2C  pid=241771
```

-------

Command:

```bash
./tools index view file://$HOME/workspace/indices \
    GCJ7F2CINZQ2WWXTS2X3BOY4UUAM4X7Q4PSKCQPI6URA56AQAUND724S,GCMY3A5MTQKSFQYMJWTNVVMSJ3GRW567UVPIP4TYZGJ2O5YGCLIPVUQ7
```

```
INFO[2022-08-29T17:06:01.406-07:00] Analyzing indices at file:///home/george/workspace/indices  pid=287475
INFO[2022-08-29T17:06:01.408-07:00] Index found: 'all/all'                        pid=287475
INFO[2022-08-29T17:06:01.408-07:00] Index found: 'all/payments'                   pid=287475
INFO[2022-08-29T17:06:01.408-07:00] Index found: 'successful/all'                 pid=287475
INFO[2022-08-29T17:06:01.409-07:00] Index found: 'successful/payments'            pid=287475
INFO[2022-08-29T17:06:01.409-07:00] Activity for account:                         account=GCJ7F2CINZQ2WWXTS2X3BOY4UUAM4X7Q4PSKCQPI6URA56AQAUND724S limit=10 pid=287475
INFO[2022-08-29T17:06:01.409-07:00]   - checkpoint 22039, ledgers [1410432, 1410496)  indices="all/all, all/payments, successful/all, successful/payments" pid=287475
INFO[2022-08-29T17:06:01.409-07:00] Summary: 1 active checkpoints, 64 possible active ledgers  pid=287475
INFO[2022-08-29T17:06:01.409-07:00] Checkpoint range: [22039, 22039]              pid=287475
INFO[2022-08-29T17:06:01.411-07:00] Index found: 'all/all'                        pid=287475
INFO[2022-08-29T17:06:01.411-07:00] Index found: 'all/payments'                   pid=287475
INFO[2022-08-29T17:06:01.411-07:00] Index found: 'successful/all'                 pid=287475
INFO[2022-08-29T17:06:01.411-07:00] Index found: 'successful/payments'            pid=287475
INFO[2022-08-29T17:06:01.411-07:00] Activity for account:                         account=GCMY3A5MTQKSFQYMJWTNVVMSJ3GRW567UVPIP4TYZGJ2O5YGCLIPVUQ7 limit=10 pid=287475
INFO[2022-08-29T17:06:01.411-07:00]   - checkpoint 22034, ledgers [1410112, 1410176)  indices="all/all, all/payments, successful/all, successful/payments" pid=287475
INFO[2022-08-29T17:06:01.411-07:00] Summary: 1 active checkpoints, 64 possible active ledgers  pid=287475
INFO[2022-08-29T17:06:01.411-07:00] Checkpoint range: [22034, 22034]              pid=287475
```

---------

Command: `./tools index stats file://$HOME/workspace/indices `
Output:

```
[... lots of output like the previous command abridged ...]
INFO[2022-08-29T15:49:21.939-07:00] === Final Summary ===                         pid=242344
INFO[2022-08-29T15:49:21.939-07:00] Path:     file:///home/george/workspace/indices  pid=242344
INFO[2022-08-29T15:49:21.939-07:00] Accounts: 24563                               pid=242344
INFO[2022-08-29T15:49:21.940-07:00] Smallest checkpoint: 22032                    pid=242344
INFO[2022-08-29T15:49:21.940-07:00] Largest  checkpoint: 22992                    pid=242344
INFO[2022-08-29T15:49:21.940-07:00] Checkpoint count:    961 (61504 possible ledgers, ~4.27 days)  pid=242344
INFO[2022-08-29T15:49:21.940-07:00] Index names: successful/payments, all/all, all/payments, successful/all  pid=242344
INFO[2022-08-29T15:49:22.122-07:00] Most active account: GCFJN22UG6IZHXKDVAJWAVEQ3NERGCRCURR2FHARNRBNLYFEQZGML4PW (961 checkpoints)  pid=287361
```

### CODE REVIEW

File: exp/lighthorizon/tools/explorer.go [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+			log.Infof("Analyzing indices at %s", path)
+
+			allCheckpoints := set.Set[uint32]{}
+			allIndexNames := set.Set[string]{}
+			accounts := showAccounts(path, 0)
+			log.Infof("Analyzing indices for %d accounts.", len(accounts))
+
+			// We want to summarize as much as possible on a Ctrl+C event, so
+			// this handles that by setting up a context that gets cancelled on
+			// SIGINT.
```
Comment (sreuland) [inline exp/lighthorizon/tools/explorer.go:62]: nice, maybe worth mentioning that it also supports second ctrl-c to hard stop immediately and not wait for summarization to wind down after first ctrl-c?

File: exp/lighthorizon/tools/explorer.go [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+			//
+			// https://pace.dev/blog/2020/02/17/repond-to-ctrl-c-interrupt-signals-gracefully-with-context-in-golang-by-mat-ryer.html
+			c := make(chan os.Signal, 1)
+			signal.Notify(c, os.Interrupt, syscall.SIGTERM, syscall.SIGINT)
+			ctx, cancel := context.WithCancel(context.Background())
```
Comment (sreuland) [inline exp/lighthorizon/tools/explorer.go:67]: just for consideration, noticed `signal.NotifyContext()` , may eliminate need for channel and select statement plumbing, saw it used here - https://millhouse.dev/posts/graceful-shutdowns-in-golang-with-signal-notify-context

Comment (Shaptic) [reply exp/lighthorizon/tools/explorer.go]: awesome tip, much cleaner in 99c7ff0b!

File: exp/lighthorizon/tools/explorer.go (hunk 1 above)
Comment (Shaptic) [reply exp/lighthorizon/tools/explorer.go:64]: good call, added a log message around this :+1:

**Pull request merged**